### ISSUE TITLE
Issue with fluide grid (percentage width columns)

### ISSUE THREAD

**Issue by 7R5N** [desandro/masonry#795]
Hi, 

I’m not sure, but I don’t understand the way the left position is calculated on layouts based on percentage widths. 

That even doesn’t work correctly in the example provided on your webpage: [http://codepen.io/desandro/pen/JFpeg]

- One Column is 20% wide
- an Item in the second column has a left value of "19.777643590381132%", or "19.79269328802039%", or "19.813567394706403%", depending on the width of the container
→ I’m not a genius on maths … but, I thought that the second column should always start at 20% instead. 

This uneven percentages lead to 1px flashing gutters, depending on the size of the container. 
@MacOSX; Safari, Chrome, Firefox …

To me, this looks like the script re-calulates to percent, what the browser formerly calculated in pixels from the percentages provided by CSS, like: 
css/percentage → browser/pixel → script/percentage

I’m aware that it’s not always possible to count with even pixels on fluid layouts: you simply can’t divide 981 pixels by 5 columns with non-comma-pixel-widths. 

But, that’s what the percentage properties are for. They tell the browser "that’s your job". And the browsers are able to dived the space in columns. And when I set the left values to 20% (40%, 60%, 80%) in the WebInspector, there are no flashing gutters. 

→ Wouldn’t it be better to get/read the CSS width percentage value to directly set the left value …?

**Comment by desandro** [comment]
Thank you for this feedback! This is an interesting idea. Internally, Masonry has always used pixels for its measurement of items and column sizes. The `percentPosition` option allows for item position to be rendered with percentages. As you noted, the output may be off due to imperfect measurement, pixel rounding, and general JavaScript decimal wonkiness.

Using percent values internally could resolve the 1 pixel gaps, but it would require re-tooling a fair amount of code to do so. I'm open to the idea, but I'd like to think on it for a bit.

**Comment by 7R5N** [comment]
Yes, maybe that’s an even better point, to _calculate_ in percentages internally – that would be more like a strategy, other than the ~fix I suggested earlier … 
After all; the big grid systems out there (Bootstrap et al) all work with percentage widths internally.

**Comment by HriBB** [comment]
Hey there. This is a repost from the incorrectly posted `masonry-docs` issue.

First of all, thanks for this great library! It works perfectly except for this very tiny little issue.

I'm using [`react-masonry-component`](https://github.com/eiriklv/react-masonry-component), but I think this is an issue with masonry itself. Not really sure though. Anyway, I'm using `percentPosition:true` with `columnWidth` sizer selector and sometimes (usually on full width) there is this extra 1px space between some images.

Here's the screenshot

![masonry](https://cloud.githubusercontent.com/assets/4412324/18324822/b7a117d0-753e-11e6-98de-5b5a766d4818.png)

So how can we help you with this?

**Comment by guicara** [comment]
I had the same problem with my image gallery. The solution, [as explained in this page](http://metafizzy.co/blog/beyonce-seamless-fluid-image-masonry/), is to change the container width (a little bit less than 100%) and increase the column width.

![masonry-1px-gap](https://cloud.githubusercontent.com/assets/2470367/20343384/70a7e660-abee-11e6-93d3-45e910b822ab.PNG)

**Comment by rockstarwebdesign** [comment]
Obnoxious issue... did anyone find a workaround?

**Comment by desandro** [comment]
There's the [seamless Masonry via Beyoncé's tumblr solution](http://metafizzy.co/blog/beyonce-seamless-fluid-image-masonry/)

**Comment by danyj** [comment]
@desandro you know I have been on this one for long and just now have found a suitable solution for my case that might also help others, 


we all have been looking at masonry items only but I think what we have failed to look at is the grids

most of us use 6 column grid for metro ( packery )  layout and in order for isotope to be perfectly aligned that grid must be equally divisible by 6 , 

example of a bad layout 

grid  1200px
side padding 25px 
leaves grid at 1150px  and we see the 1px gaps since 1150 / 6  = 191.6666666666667

![screenshot_8](https://cloud.githubusercontent.com/assets/2340172/21468320/544ab6c4-ca0b-11e6-8799-43df1fc36b7c.png)


now increase the padding to 30px  ( 1140 / 6 = 190)  and you have perfect layout. 


![screenshot_9](https://cloud.githubusercontent.com/assets/2340172/21468324/810504da-ca0b-11e6-8d25-da1eefc08438.png)


That is on page load , but if you nail the resize to a number that is not evenly  divisible by 6 you see the bug again .


since I am working on a theme that will let user change the grid width and padding's I had to find something that would work in any case , so I came up with this that runs on page load and resize 

this method calculates the difference between the holding container and the total size of fitting items, 
the difference is than converted to percentage of holding container and that percentage is applied as width to the isotope grid. 



```
	ThzAdjustGrid: function (){
		
		var self 	= this;
		
		var $container_w 	= $(self.element).width(); // holds isotope container
		var $items_w		= Math.floor( $container_w / 6 ) * 6; // total possible width
		var $grid_pull  	= parseInt( self.grid.css('margin-left')); // grid side pull
		var $diff 		= $container_w - $items_w + $grid_pull; // the difference 
		var $perc 		= 100 - ( $diff / $container_w * 100 ); // percent of holding container
		
		self.grid.css('width',$perc +'%');
	
	},


```

I hope you dont mind the plugin method ( you get the idea ) 

in my case it works on any grid padding or grid size + on resize.

If you have better idea please do let me know.

**Comment by rockstarwebdesign** [comment]
See here: https://codepen.io/desandro/pen/myBxD

This still shows issues. Some images overlap others by 1 pixel.

**Comment by hasinhayder** [comment]
I have found a nice fix for this issue. Works 100% in all browsers. I will push it tonight :)

**Comment by tiagogon** [comment]
@hasinhayder where is the fix? :)

**Comment by zamson** [comment]
@hasinhayder please share your fix. This is driving me nuts 😞

**Comment by hasinhayder** [comment]
@zamson @tiagogon - the problem occurs because of using parseInt() in several places where it should be parseFloat()

if you are using the masonry.pkgd.js or masonry.pkgd.min.js packaged version (minified or not minified), the quickest fix is to open your js file and replace all parseInt with parseFloat. Nothing will break and your problem will be fixed right away.

**Comment by hasinhayder** [comment]
@zamson no drawbacks buddy :)

**Comment by desandro** [comment]
[Outlayer](https://github.com/metafizzy/outlayer) v2.1.1 was released last year, which uses float values. I'll be updating Masonry with this change in its next release.

**Comment by hasinhayder** [comment]
@desandro good to see that it's finally happening :)

**Comment by desandro** [comment]
Outlayer v2.1.1 has been added to masonry.pkgd.js in Masonry [v4.2.2](https://github.com/desandro/masonry/releases/tag/v4.2.2)
